### ISSUE TITLE
Allow tasks to be rerun without persisted results when a flow is resumed

### ISSUE THREAD

**Issue by masonmenges** [PrefectHQ/prefect#8360]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Once a flow run has been paused if results are not persisted then upon resuming the flow will fail with the following error 

`
  raise MissingResult(
prefect.exceptions.MissingResult: State data is missing. Typically, this occurs when result persistence is disabled and the state has been retrieved from the API.
`

MRE, requires a deployment with rescheduling enabled.

```
from prefect import task, flow, pause_flow_run
from prefect.context import get_run_context
from random import randint

@task()
def some_taska():
    print("perform some check")

    random_number = randint(0, 3)

    return random_number

@task
def some_taskb():
    print("resumed_successfully")

@flow
def paused_flow():

    task_num = some_taska()
    print(task_num)

    flow_run_id = get_run_context().flow_run.id

    if task_num >= 2:
        pause_flow_run(flow_run_id=flow_run_id)

    some_taskb()


if __name__ == "__main__":
    paused_flow()
    
```

### Describe the proposed behavior

Upon resuming a flow if no results data has been persisted on a prefect should rerun the task instead of failing the flow.

### Example Use

Often user want to check outside sources to determine before completing some of their work flows, i.e. they may be waiting for some external data to be available on an unknown time schedule, in this scenario it's beneficial to be able to run a task verifying the existence of said data and then pausing/continuing with the flow depending on the outcome of the task. If results are persisted then the flow never completes in this scenario as the upstream task will always return the same output.

**Comment by anticorrelator** [comment]
Hi @masonmenges! By supplying a `flow_run_id` to the `pause_flow_run` utility, you're performing an *out of process* pause. For example, you'll be able to remotely pause a flow, even if it's being run by a different agent using that syntax. In order to do that, you'll need a deployed flow that persists results.

For the pause behavior you're asking for, simply call the `pause_flow_run` utility without that argument from inside whatever flow you're trying to pause. That will suspend execution until it's resumed: either via the `resume_flow_run` utility or the UI.

The following example should work for you:

```python
from prefect import task, flow, pause_flow_run
from random import randint

@task()
def some_taska():
    print("perform some check")

    random_number = randint(0, 3)

    return random_number

@task
def some_taskb():
    print("resumed_successfully")

@flow
def paused_flow():

    task_num = some_taska()
    print(task_num)

    if task_num >= 2:
        pause_flow_run()

    some_taskb()


if __name__ == "__main__":
    paused_flow()
```

**Comment by jbnitorum** [comment]
@anticorrelator I do think not this is correct.  Previously I had been trying to resume a paused without its results being persisted and the flow was failing as described here.  Now I see the flow complete even when results are explicitly **not** persisted.  That's great, but the desired result is the for the tasks without persisted results to rerun on resumption, that isn't happening here.  This flow should get a new random number from some_taska after it is resumed.  In process vs out of process wasn't the issue in question here.  If you read the example use case you might better understand.  The goal here is that tasks without persisted results will re run when the flow is resumed.

The flow_run_id argument should have been omitted from the example as its not relevant to the heart of the matter since the flow is being paused in process.

Please reopen this ticket.

@EmilRex @masonmenges 

```
from prefect import task, flow, pause_flow_run
from random import randint

@task(persist_results=False)
def some_taska():
    print("perform some check")

    random_number = randint(0, 3)

    return random_number

@task
def some_taskb():
    print("resumed_successfully")

@flow
def paused_flow():

    task_num = some_taska()
    print(task_num)

    if task_num >= 2:
        pause_flow_run()

    some_taskb()


if __name__ == "__main__":
    paused_flow()
```

**Comment by madkinsz** [comment]
@anticorrelator for some more context, we previously discussed allowing a task run in a final state to run again if the result is not persisted and the flow is restarted

**Comment by cicdw** [comment]
Hi @jbnitorum I have a question - what I'm interpreting from your issue is that you want to do the following:
- pause a flow run and have it turn off the running process
- user manually decides to resume the flow run
- you want the entire flow function to rerun from the beginning

Assuming that interpretation is correct, could you explain why you need this behavior in Pause / Resume?  That seems much easier to accomplish with something like:

```python
@flow
def paused_flow():

    task_num = some_taska()
    print(task_num)

    if task_num >= 2:
        return # could also include a specific state here that you want the run to end in

    some_taskb()
```

And then the "re-run" is simply a new run of the flow. I don't see a compelling reason for this to be implemented as a pause / resume feature given that it's already equivalent to a complete rerun.

**Comment by jbnitorum** [comment]
@cicdw As I understood it now the flow does re-run from the beginning and the appearance of being resumed mid-flow is really just achieved by way of task result persistence (tasks with results don't re-run).  If my understanding is incorrect let me know.

I keep circling back to this but we are a v1 client trying to fulfill our obligation to migrate to v2 without completely redesigning how we interact with Prefect.  We have hundreds of flows configured using a pause/resume structure in v1 and the combination of two factors is making making it a giant lift for us because there isn't feature parity around pause/resume between v1 and v2.  **At the end of the day all we are trying to do is to do in v2 what we are already doing in v1.**  So please... understand my frustration and persistence on the issue.  The two factors are:

- Inability to pause tasks.  This is how v1 works, would be easiest way to make the transition.  Exact feature parity.  Requested [here](https://github.com/PrefectHQ/prefect/issues/8442).
- Inability to have some task results persist and others not persist in a flow with a pause.  This is what we are talking about in this post.  This allows for similar behavior to v1.  In v1 your task calling a pause never hits a terminal state so it reruns when resumed.  You get to a similar place here by just explicitly having that same task re-run via persist_result=False.  That is the core of this issue in this post.

When using remote infrastructure you are met with this error when attempting to resume a flow with tasks where results were not persisted:

_MissingResult: State data is missing. Typically, this occurs when result persistence is disabled and the state has been retrieved from the API._

_**My request on this post is that instead of this error, just have the task without results re run.**_


The example we are using here is oversimplified to illustrate what is being requested from a technical standpoint (rerunning of some tasks on pause resumption but not others).  Sure you could just rerun the flow and that sounds fine in vacuum with no context but our flows in reality are more complex than this simplified example.  Here's a more complicated example written in layman's terms that is more similar to what we are actually doing.  You could make the argument from an engineering standpoint that this would be more effective as 3 separate flows but when you do that you lose a ton of observability benefits that are core to Prefect's value proposition.  Engineering aside, imagine telling the restaurant manager that he needs to check 3 places to see if the food got made instead of 1.  We need to think of stakeholders other than engineers.


```python
@task(name="Get shopping list", persist_result=True)
get_shopping_list(recipe):
    shop_list = getting_shopping_list_from_website(recipe)
    email_shopping_list_to_shopper(shop_list)
    place_item_on_menu(recipe)
    return shop_list

@task(name="Check pantry", persist_result=False)  ####<<<<<<<result is not persisted
check_pantry(shop_list):
    ### this checks our pantry to see if the shopper bought all the ingredients and placed them in the pantry
    pantry_status = does_pantry_have_ingredients(shop_list)
    if pantry_status == False:
        #### flow pauses here almost always when run initially as the shopper can’t buy and stock the pantry instantaneously... however we do have instances where all the items happen to already be in the pantry so we can't assume a pause, we need to run this test
        task_context = get_run_context()
        pause_flow_run(flow_run_id=task_context.task_run.flow_run_id)
        ### in 1.0 this is ----   raise PAUSE('Items not in pantry')
    return None

@task(name="Get ingredients", persist_result=True)
Move_ingredients_to_kitchen(shop_list):
    carry_ingredients(shop_list)
    return None

@task(name="Start cooking", persist_result=True)
start_cooking():
    throw_every_thing_in_pot()
    turn_on_stove()
    return None

@task(name="Check temperature of food", persist_result=False)  ####<<<<<<<result is not persisted
check_temp(recipe):
    temp = use_virtual_thermometer()
    required_temp = get_required_temp(recipe)
    if temp < required_temp:
        #### flow pauses here almost always when run initially as the food generally doesn't heat instantaneously... 
        #### however we do have instances where its a cold dish and doesn't need to cook so we can't assume a pause, we need to run this test
        task_context = get_run_context()
        pause_flow_run(flow_run_id=task_context.task_run.flow_run_id)
        ### in 1.0 this is ----   raise PAUSE('Still cooking')
    return

@task(name+"Bring food to diner",  persist_result=True)
bring_out_food():
    carry_to_diner()
                
@task(name="Send out Bill", persis_result=True)
send_out_bill():
    send_out_bill()

@flow(name="Cook Recipe")
Cook_recipe(recipe=’Chili’):
    shop_list = get_shopping_list(recipe)
    pantry_checked = check_pantry(shop_list) 
    ##### flow pauses on task above if items aren’t in pantry.  When personal shopper placed item in pantry they hit a button that runs a script that looks for an instance of the Cook Recipe flow in a paused state and resumes that flow via the API  
    ##### That runs the task above again as results were not persisted.  If they forgot to buy something the flow will pause again as we wont have ingredients we need.
    ingredients_in_kitchen = move_ingredients_to_kitchen(shop_list, wait_for=[pantry_checked])
    cooking_started = start_cooking(wait_for=[ingredients_in_kitchen])
    temperature_checked = check_temp(recipe, wait_for=[cooking_started])
    ##### flow pauses on task above if the food isn't hot enough.  When the chef thinks the food is ready they hit a button that runs a script that looks for an instance of the Cook Recipe flow in a paused state and resumes that flow via the API.  
    #### That runs the task above again as results were not persisted.  If the food still isn't hot enough it pauses again.
    #### when we resume here it would be ok for check_pantry to rerun (remember results aren't persisted there)  
    #### I know it doesn't make sense since the ingredients were removed from pantry, but bear with me... i made this whole thing up so its relatable/understandable.  Don't think of this as a constraint.
    food_served = bring_out_food(wait_for=[temperature_checked])
    send_out_bill(wait_for=[food_served])
```

**Comment by cicdw** [comment]
@jbnitorum thanks for the additional info, I'm still processing it; in the meantime would you mind sharing why `some_taska` is coded as a task instead of pure flow logic?  Is it for observability in the UI, retries, or something else?

**Comment by jbnitorum** [comment]
There’s definitely a little “because we did it that way in v1” honestly.  The biggest reason is UI observability.  Expanding on that, my thought has been given flows that could pause at multiple points (as in the longer example) it gets harder to follow what is going on if the results driving the pause logic can’t be observed as tasks in the UI.  It’s harder to tell at what point the pause happened…. Although maybe there is an alternative way to do this like updating some metadata around that state of the pause like the name or message I.e. “flow paused after task x”

It’s not lost on me that we could probably just code this into the flow but I was kind of hoping for a cleaner solution like we had in v1.  This feels like we are moving backwards.  

There are a lot of uses for the rerunning of task within a flow with pausing related to checking if a resource is an available.  

You could check for a file 3 times using retries and a delay.  If the third fails you want to pause so you don’t keep running infrastructure unnecessarily.  On resume the same checking task runs again to ensure the resource is there.  This is a better design than just assuming that whatever/whoever resumed the flow ensured the required resource is present.

ideally if a task could be rerun you could have this:

Check gas gauge -> gauge says empty -> pause to refuel -> resume -> check gas gauge again -> pause again if gas still empty and repeat checking process on resumption OR drive if tank is full


Currently it works like this:

Check gas gauge -> gauge says empty -> pause to refuel -> resume -> drive without verifying tank is full… hopefully it is



To put this in the most succinct way possible:
**if there is a task result that determines if a flow should pause there should be an option to rerun that task when the flow is resumed**

@kielosz has expressed similar needs on another thread

**Comment by anticorrelator** [comment]
@jbnitorum I understand the use case and I do believe that high level pattern is an important one—determining if execution if "ready" to proceed is helpful in many contexts.

With the particular change proposed here, we risk weakening some of the abstractions of Prefect 2, such as the consistency of downstream tasks. Because the task graph is "discovered" as the flow code is executed and Prefect will attempt to not run the same task twice, a flag to always "re-execute" tasks on retries/resumes comes with *many* caveats. You can imagine enabling this flag for a single task with downstream dependencies can lead to many unintuitive situations: if the task in question is not a pure function, downstream dependencies may suddenly return values that are inconsistent for the run, or the flow may fail to branch properly downstream.

I also worry about strongly coupling the concept of persisting results to the "re-execution" properties of a task. The connection between the two will be further muddied when we introduce the concept of Artifacts as well.

I don't think a feature to force re-execution is necessarily off the table, it's just a change that has a lot of implications we'd like to consider more carefully. Pause/resume functionality should soon be available as an automation (cc @WillRaphaelson), and I believe observable, event-driven pausing/resuming of a flow is the most powerful way to do it (and also the solution I had in mind when considering patterns where a flow must wait on a health check to run).

Another thought as far as observability goes: we hope to enable the creation of Artifacts from tasks and flows soon. Perhaps, if you decide to move the pausing logic into the flow you can preserve the ready-state of the flow as an artifact each time?

**Comment by jbnitorum** [comment]
Artifacts could be a way to observe the reason for a flow’s pause.  Could you explain the pause/resume automation and how it would work in this workflow?

**Comment by anticorrelator** [comment]
One way would to set up a deployment that runs a flow that loops until some property of your system is ready, at which point it will complete, something like:

```python
from anyio import move_on_after
from prefect import flow

@flow
async def ready_checker():
    one_hour = 60 * 60
    with move_on_after(one_hour):
        while True:
            await anyio.sleep(10)
            if is_ready():
                return  # flow will complete
    raise NotReadyError()  # otherwise flow fails
```

Soon, we'll have better way to generate custom events via the API, but that ready checker flow will fire either a completion or failure event on termination. The primary flow you want to run will simply run like this:

```python
@flow
def my_flow():
    start_stuff()
    run_deployment(ready_checker)
    pause_flow_run(reschedule=True)
    finish_stuff()
```

Coupled with an automation that will resume `my_flow` whenever the `ready_checker` deployment enters a `Completed` state.

**Comment by jbnitorum** [comment]
@anticorrelator yeah so the main issue here that pause/resume solves for is cost.  When you have hundreds of flows running that all wait for several hours on resources you end up using a ton of compute time that is effectively doing nothing.  For us the compute cost related to waiting would probably be 100x what it is actually running operations.

I get what you are saying conceptually, but I'm having trouble wrapping my head around how we do this at scale.  We have hundreds of flows using pause resume that each wait for their own independent resources on a run by run basis.  So ready_checker couldn't be scheduled.  Each flow run would need to pass params to its own ready_checker run and then the automation would need to know which flow run to resume somehow.  Only way I can think of this working is passing a flow_run_id as a parameter to a third deployment tied to the automation.  We end up with so many extra pieces to run a single process.  More points of failure. Less observability because the pieces are in 3 places instead of 1 (flow, ready_checker deployment, automation deployment).  We achieve all of this in v1 just by being able to pause a task if condition is not met, it exits without a result, then resume a task and its code runs from the top, if the condition is met it completes, if its not met it pauses again.  We see everything in one place, no extra pieces.

Going back to your previous post about the feature proposed on this thread presenting issues for downstream tasks.  Might the pausing of tasks (exits without result) make for a better solution?

I keep harping on this but we had a discussion almost a year ago when we first built all of this in v1 and asked the question of 'will this pose a problem in v2?' and the response was consistently that the plan was the full support v1 feature set in v2.  At this point I think we have a situation where despite asking the right questions in advance we are going to have a massive lift on our end to migrate to v2 and we end up with something that is actually less useful to us.  While I get under the hood v2 is a lot different and it may not be able to function in exactly the same way, please understand that there was a lot of ink, paper, and promises made that preceded this thread.  I'm trying to work with you guys on this and I've put quite a bit of time into this but part of me just wants to say "you need build this feature because you said you would to get us to sign a contract."

**Comment by anticorrelator** [comment]
@jbnitorum if you ran a scheduled deployment that simply checked once every 10 minutes for a ready state and triggered an automation off of that to resume paused flows, would that not drastically remove busy-waiting?

**Comment by jbnitorum** [comment]
@anticorrelator Yeah I guess that would work, but that comes at the cost of running a new ready_checker deployment 12 times if the wait is 2 hours.  I suppose we could filter out the ready_checker runs with a tag so it doesn't muddy our high level view of whats going on in our workspace.  The whole thing just feels too complicated to me.  Too many moving parts to do something simple.

It sounds like our only option to have simplicity is putting all the code that decides to pause the flow into the flow itself rather than a task.  Do you see any issue with the below?  This flow has two pause/approvals with a task run in between.  The one thing I'm not quite clear about this with regards to v1 vs v2 is execution order.  In v1 the flow was just sort of a blueprint where as in v2 its actual running code.  In v1 you had to sort of map out the order using upstream_tasks etc.

So in this below will check_for_resource("resource2") only run after task_b has completed?  Do I even need the wait_for on task_d if I'm not using async?

```python

@flow()
    task_a()

    is_resource1_present = check_for_resource("resource1") #### not a task
    if  not is_resouce1_present:
        save_pause_location("After task a, before task b")  #### not a task
        pause_flow_run(reschedule=True, key=unique_timestamp())
    else:
        task_b()
        is_resource2_present = check_for_resource("resource2") #### not a task, will this only executed after task_b has completed?
        if not is_resource2_present:
            save_pause_location("After task b, before task c")  #### not a task
            pause_flow_run(reschedule=True, key=unique_timestamp())
        else:
            task_c_result = task_c()

    task_d(wait_for=[task_c_result])

```
